### PR TITLE
[Serve] Remove global async thread: takes 2

### PR DESCRIPTION
**Pull request by simon-mo**
## Why are these changes needed?
Superseding #15379

Original comments:
This PR removes the background thread by using the _on_completed callback on the tracker_ref in ReplicaSet.

I think this is a nice state to be at for Serve v2 API. There are still difference between SyncHandle and AsyncHandle. But the underlying mechanism is much cleaner now and both uses the same callback based API.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/router.py [hunk 1]
```diff
@@ -76,17 +86,14 @@ def __init__(
         # the same node.
         self.replica_iterator = itertools.cycle(self.in_flight_queries.keys())
 
-        # Used to unblock this replica set waiting for free replicas. A newly
-        # added replica or updated max_concurrent_queries value means the
-        # query that waits on a free replica might be unblocked on.
-        self.config_updated_event = asyncio.Event(loop=event_loop)
+        self.pending_requests: Deque[PendingQuery] = deque()
```
Comment (edoakes) [inline python/ray/serve/router.py:89]: inconsistent naming here. let's pick either "query" or "request" and stick with it

File: python/ray/serve/router.py [hunk 2]
```diff
@@ -146,64 +164,41 @@ def _try_assign_replica(self, query: Query) -> Optional[ray.ObjectRef]:
...
+            return True
+        return False
+
+    def _on_query_tracker_completed(self, _tracker_result, *,
+                                    replica_handle: ActorHandle,
+                                    tracker_ref: ray.ObjectRef):
+        tracker_refs: Set[ray.ObjectRef] = self.in_flight_queries.get(
+            replica_handle, {})
+        tracker_refs.discard(tracker_ref)
+        self._try_send_pending_queries()
```
Comment (edoakes) [inline python/ray/serve/router.py:188]: what's the behavior when the actor request fails? do we need to handle it differently at all?

**Review by edoakes — APPROVED** [review]

**Comment by simon-mo** [comment]
benchmarked throughput: we are at 1934 qps vs the master 2142 qps

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Pull request closed without merging**